### PR TITLE
[Enhancement] Reduce head object requests in lake tablet GC part 2/2

### PR DESCRIPTION
**Pull request by sduzh**
This is the last one of the PR stack to reduce head object requests in lake tablet GC:
- Support iterate_dir2() in StarletFileSystem, which may reduce the head object requests in lake tablet GC
- Refactor: update FileSystem::iterate_dir2() interface signature

## What type of PR is this：
- [ ] BugFix
- [ ] Feature
- [x] Enhancement
- [ ] Refactor
- [ ] UT
- [ ] Doc
- [ ] Tool

## Which issues of this PR fixes ：

Fixes #

## Checklist:

- [ ] I have added test cases for my bug fix or my new feature
- [ ] This pr will affect users' behaviors
- [ ] This pr needs user documentation (for new or modified features or behaviors)
  - [ ] I have added documentation for my new feature or new function

## Bugfix cherry-pick branch check:
- [x] I have checked the version labels which the pr will be auto backported to target branch
  - [x] 3.0
  - [ ] 2.5
  - [ ] 2.4
  - [ ] 2.3

### CODE REVIEW

**Review by imay — DISMISSED** [review]

**Comment by sduzh** [comment]
run starrocks_admit_test

File: be/src/runtime/snapshot_loader.cpp [hunk 1]
```diff
@@ -803,27 +803,27 @@ Status SnapshotLoader::_get_existing_files_from_remote_without_broker(const std:
...
-    st.update(fs->iterate_dir2(remote_path, [&](std::string_view name, const FileMeta& meta) {
-        if (UNLIKELY(!meta.has_is_dir())) {
+    st.update(fs->iterate_dir2(remote_path, [&](DirEntry entry) {
+        if (UNLIKELY(!entry.is_dir.has_value())) {
             st.update(Status::InternalError("Unable to recognize the file type"));
             return false;
         }
         file_num++;
-        if (meta.is_dir()) {
+        if (!entry.is_dir.value()) {
```
Comment (wyb) [inline be/src/runtime/snapshot_loader.cpp:812]:
```suggestion
        if (entry.is_dir.value()) {
```

**Review by wyb — DISMISSED** [review]

**Comment by sduzh** [comment]
run starrocks_admit_test

**Comment by sduzh** [comment]
run all

**Review by wyb — APPROVED** [review]

**Review by gengjun-git — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
@Mergifyio backport branch-3.0

**Pull request merged**